warehouse/query-core/src/main/java/datawave/query/iterator/waitwindow/WaitWindowObserver.java

```diff
@@ -0,0 +1,417 @@
+                if (origColQual.isEmpty()) {
+                    colQual = marker;
+                } else if (yieldToBeginning) {
+                    colQual = new Text(marker.toString() + "\0" + origColQual);
+                } else {
+                    colQual = new Text(origColQual + marker.toString());
+                }
+            }
+            newKey = new Key(key.getRow(), key.getColumnFamily(), colQual);
+        }
+        return ensureYieldKeyAfterRangeStart(newKey);
+    }
+
+    // When the current seekRange is non-inclusive, we can not return the startKey of the range
+    // as a yield key. Instead, we have to return the following key.
+    private Key ensureYieldKeyAfterRangeStart(Key key) {
+        if (!this.seekRange.isStartKeyInclusive()) {
+            Key seekStartKey = this.seekRange.getStartKey();
+            boolean isShardKey = isShardKey(seekStartKey);
+            if (isShardKey && key.compareTo(seekStartKey, PartialKey.ROW_COLFAM) <= 0) {
+                // shard range
+                return seekStartKey.followingKey(PartialKey.ROW_COLFAM);
+            }
+            if (!isShardKey && key.compareTo(seekStartKey, PartialKey.ROW_COLFAM_COLQUAL) <= 0) {
+                // document range
+                return seekStartKey.followingKey(PartialKey.ROW_COLFAM_COLQUAL);
+            }
+        }
+        return key;
+    }
+
+    private Key yieldKeyAfterOverrun(Key key) {
+        Text row = key.getRow();
+        Text colFam = key.getColumnFamily();
+        Text colQual = key.getColumnQualifier();
+        if (hasBeginMarker(key)) {
+            if (hasBeginMarker(colFam)) {
+                colFam = new Text(colFam.toString().replace(YIELD_AT_BEGIN_STR, YIELD_AT_BEGIN_STR + "\0"));
+            } else {
+                colQual = new Text(colQual.toString().replace(YIELD_AT_BEGIN_STR, YIELD_AT_BEGIN_STR + "\0"));
+            }
+        } else if (hasEndMarker(key)) {
+            if (hasEndMarker(colFam)) {
+                colFam = new Text(colFam.toString().replace(YIELD_AT_END_STR, YIELD_AT_END_STR + "\0"));
+            } else {
+                colQual = new Text(colQual.toString().replace(YIELD_AT_END_STR, YIELD_AT_END_STR + "\0"));
+            }
+        }
+        return new Key(row, colFam, colQual);
+    }
+
+    public void setYieldKey(Key yieldKey) {
+        // only do this once
+        if (this.trackingSpan != null && this.yieldKey == null) {
+            this.trackingSpan.yield();
+        }
+        // we can't return a key and then yield at the same key
+        this.yieldKey = yieldKeyAfterOverrun(yieldKey);
+    }
+
+    public boolean isReadyToYield() {
+        return readyToYield && yieldKey != null;
+    }
+
+    public Key getYieldKey() {
+        return yieldKey;
+    }
+
+    static public boolean isShardKey(Key key) {
+        Text colFam = key.getColumnFamily();
+        return colFam.equals(new Text()) || hasBeginMarker(colFam);
```

Comment:
Why is this check sufficient?  It is not obvious that a begin markiner in the colFam is necessarily a shard key.

Reply:
Comments added.  Document-specific keys have YIELD_AT_BEGIN or YIELD_AT_END in the colQual, whicle shard keys (non-document) have these markers in the colFam.